src/QsCompiler/Core/Transformations/TypeTransformation.fs

```diff
@@ -17,22 +17,46 @@ open Microsoft.Quantum.QsCompiler.Transformations.Core.Utils
 type private ExpressionType = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>
 
 
-type TypeTransformationBase(options: TransformationOptions) =
+type TypeTransformationBase internal (options: TransformationOptions, _internal_) =
+
+    let missingTransformation name _ =
+        new InvalidOperationException(sprintf "No %s transformation has been specified." name) |> raise
+
     let Node = if options.Rebuild then Fold else Walk
 
-    new() = TypeTransformationBase TransformationOptions.Default
+    member val internal CommonTransformationItemsHandle = missingTransformation "common items" with get, set
+
+    member this.Common = this.CommonTransformationItemsHandle()
+
+    internal new(getCommonItems: unit -> CommonTransformationItems, options: TransformationOptions) as this =
+        new TypeTransformationBase(options, "_internal_")
+        then this.CommonTransformationItemsHandle <- getCommonItems
 
+    new(options: TransformationOptions) as this =
+        new TypeTransformationBase(options, "_internal_")
+        then
+            let commonItems = new CommonTransformationItems()
+            this.CommonTransformationItemsHandle <- fun _ -> commonItems
+
+    new() = TypeTransformationBase TransformationOptions.Default
```

Comment:
I think this can be simplified a lot. You don't need `missingTransformation` or a constructor with an unused `_internal_` parameter if you make the primary constructor take in a `getCommonItems`.

```fsharp
type TypeTransformationBase internal (options, getCommonItems) =
    static let createCommonGetter () =
        let common = CommonTransformationItems()
        fun () -> common

    let Node = if options.Rebuild then Fold else Walk

    member val internal CommonTransformationItemsHandle = getCommonItems with get, set

    member this.Common = this.CommonTransformationItemsHandle()

    new(options) = TypeTransformationBase(options, createCommonGetter ())

    new() = TypeTransformationBase TransformationOptions.Default
```

Reply:
It just seemed so inconsistent, and I got rid of having independent initialization for all subtransformations. This seems all in all much simpler, and I was just bothered by one constructor looking different than for the ones for the rest of the transformations.

Reply:
Sorry, I'm not sure what you mean? In terms of consistency, I think that all of the transformation constructors could be cleaned up in the same way to avoid the `_internal_` parameter hack.

Reply:
Looking through the transformations as they are before this PR, I am also not seeing the need for the "_internal_" constructors when they are just called by, what I would consider, the proper constructors. Though I am also not sure what the purpose of the `getCommonItems` parameter is either.

Reply:
For example with the TypeTrasformation:
```fsharp
and TypeTransformation<'T> internal (options, _internal_) =
    inherit TypeTransformationBase(options)
    let mutable _Transformation : SyntaxTreeTransformation<'T> option = None // will be set to a suitable Some value once construction is complete

    /// Handle to the parent SyntaxTreeTransformation.
    /// This handle is always safe to access and will be set to a suitable value
    /// even if no parent transformation has been specified upon construction.
    member this.Transformation
        with get () = _Transformation.Value
        and private set value = _Transformation <- Some value

    member this.SharedState = this.Transformation.SharedState

    new(parentTransformation: SyntaxTreeTransformation<'T>, options: TransformationOptions) as this =
        new TypeTransformation<'T>(options, "_internal_")
        then this.Transformation <- parentTransformation

    new(sharedState: 'T, options: TransformationOptions) as this =
        TypeTransformation<'T>(options, "_internal_")
        then
            this.Transformation <- new SyntaxTreeTransformation<'T>(sharedState, options)
            this.Transformation.Types <- this

    new(parentTransformation: SyntaxTreeTransformation<'T>) =
        new TypeTransformation<'T>(parentTransformation, TransformationOptions.Default)

    new(sharedState: 'T) = new TypeTransformation<'T>(sharedState, TransformationOptions.Default)
```
I believe could be rewritten as:
```fsharp
and TypeTransformation<'T> internal (parentTransformation: SyntaxTreeTransformation<'T>, options: TransformationOptions) =
    inherit TypeTransformationBase(options)
    let mutable _Transformation : SyntaxTreeTransformation<'T> = parentTransformation

    /// Handle to the parent SyntaxTreeTransformation.
    /// This handle is always safe to access and will be set to a suitable value
    /// even if no parent transformation has been specified upon construction.
    member this.Transformation
        with get () = _Transformation
        and private set value = _Transformation <- value

    member this.SharedState = this.Transformation.SharedState

    new(sharedState: 'T, options: TransformationOptions) as this =
        TypeTransformation<'T>(new SyntaxTreeTransformation<'T>(sharedState, options), options)
        then
            this.Transformation.Types <- this

    new(parentTransformation: SyntaxTreeTransformation<'T>) =
        new TypeTransformation<'T>(parentTransformation, TransformationOptions.Default)

    new(sharedState: 'T) = new TypeTransformation<'T>(sharedState, TransformationOptions.Default)
```
eliminating the `_internal_` constructor. Though, here again I'm not sure where the new `getCommonItems` fits into this formulation. This is just my take on the pre-existing framework.

Reply:
Sarah's comment applied to a different file, where part of the reason for the setup is that I wanted to avoid having to put all transformations into the same file. @ScottCarda-MS Your edit I believe is intended for the SyntaxTreeTransformation.fs file where the recursive type definition is not a problem (here we assemble it all, after all). Try to apply the same change to the rest of the transformations in that file (along with the corresponding edit to the constructor calls in the SyntaxTreeTransformation), and you should see the unit test failures - unless I am misinterpreting what you had in mind?

Reply:
Alright, I made a branch where I removed the `_internal_` constructors from the transformation framework: [here](https://github.com/microsoft/qsharp-compiler/compare/sccarda/RemoveInternalConstructors)
@bettinaheim you are right that a lot of the unit tests fail after this, although none of the failures really tell me why. I'm not sure what the recursive problem is, but there is clearly an issue with doing this. Since it is not the focus of this PR, maybe you could catch me up on what the root issue is offline.